### PR TITLE
[0255] Enchance body information

### LINKED ISSUE #1234: [SDL 0255] Enhance BodyInformation vehicle data
Proposal: [Enhance BodyInformation vehicle data](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0255-Enhance-BodyInformation-vehicle-data.md)

> This proposal is to enhance BodyInformation with more params to get access to more door related signals.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/847

### PR DESCRIPTION
**Pull request by kostyaBoss**
Fixes #1234 

This PR is **[ready]** for review.

### Related PR's
[Core](https://github.com/smartdevicelink/sdl_core/pull/3584)
[HMI](https://github.com/smartdevicelink/sdl_hmi/pull/466)

### Risk
This PR makes **[minor]** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).
- [x] I have tested Android, Java SE, and Java EE

#### Unit Tests
Added unit tests for the new structures

### Summary
Changes according to #1234  

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by kostyaBoss** [comment]
@santanamk ready for the Ford review

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/GateStatusTest.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
+package com.smartdevicelink.test.rpc.datatypes;
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/GateStatusTest.java:1]: @kostyaBoss I would rename this file to `GateStatusTests.java`.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/GateStatusTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Hashtable;
+import java.util.Iterator;
+
+public class GateStatusTest extends TestCase {
+    private GateStatus msg;
+
+    @Override
+    public void setUp() {
+        msg = new GateStatus(TestValues.GENERAL_GRID, TestValues.GENERAL_DOOR_STATUS_TYPE);
+        msg.setStatus(TestValues.GENERAL_DOOR_STATUS_TYPE);
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/GateStatusTest.java:26]: @kostyaBoss You can remove line 26  (`msg.setStatus(...)`) since you are already setting the status using the `DoorStatus` constructor.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/GateStatusTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    public void testRpcValues() {
+        // Test Values
+        Grid location = msg.getLocation();
+        DoorStatusType status = msg.getStatus();
+
+        // Valid Tests
+        assertTrue(Validator.validateGrid(msg.getLocation(), location));
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/GateStatusTest.java:35]: @kostyaBoss I would delete line 35 (`assertTrue(Validator.validateGrid(...))` and replace it with 

`assertEquals(TestValues.MATCH, TestValues.GENERAL_GRID, location);`

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/RoofStatusTests.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        msg.setState(TestValues.GENERAL_WINDOW_STATE);
+    }
+
+    public void testRpcValues() {
+        // Test Values
+        Grid location = msg.getLocation();
+        DoorStatusType status = msg.getStatus();
+
+        // Valid Tests
+        assertTrue(Validator.validateGrid(msg.getLocation(), location));
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/RoofStatusTests.java:36]: @kostyaBoss I would delete line 36 (assertTrue(Validator.validateGrid(...)`) and replace it with

`assertEquals(TestValues.MATCH, TestValues.GENERAL_GRID, location);`

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/RoofStatusTests.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        assertEquals(TestValues.MATCH, TestValues.GENERAL_DOOR_STATUS_TYPE, status);
+    }
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/RoofStatusTests.java:38]: @kostyaBoss Below line 37 I would add this line `assertEquals(TestValues.MATCH, TestValues.GENERAL_WINDOW_STATE, windowState);`

File: base/src/main/java/com/smartdevicelink/proxy/rpc/BodyInformation.java [hunk 6]
```diff
@@ -1,215 +1,423 @@
-/*
- * Copyright (c) 2017 - 2019, SmartDeviceLink Consortium, Inc.
```
Comment (santhanamk) [inline base/src/main/java/com/smartdevicelink/proxy/rpc/BodyInformation.java:2]: @kostyaBoss It looks like you used the code generator tool to re-generate the entire file.  Can you just merge in the changes that are specific to this PR (in the HTML these would be the parameters from `driverDoorAjar` to `roofStatuses`)?

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/responses/GetVehicleDataResponseTests.java [hunk 7]
```diff
@@ -202,6 +202,9 @@ public void testJson() {
             bodyInformationObj.put(BodyInformation.KEY_IGNITION_STABLE_STATUS, VehicleDataHelper.BODY_INFORMATION_IGNITION_STATUS);
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/responses/GetVehicleDataResponseTests.java:202]: @kostyaBoss I think you can revert these changes, as vehicle data does not need to be changed for this PR.  The corresponding PR for the JavaScript suite (https://github.com/smartdevicelink/sdl_javascript_suite/pull/354/files) did not make changes to the vehicle data.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/notifications/OnVehicleDataTests.java [hunk 8]
```diff
@@ -320,6 +320,9 @@ public void testJson() {
             bodyInformationObj.put(BodyInformation.KEY_PARK_BRAKE_ACTIVE, VehicleDataHelper.BODY_INFORMATION_PARK_BRAKE);
             bodyInformationObj.put(BodyInformation.KEY_IGNITION_STABLE_STATUS, VehicleDataHelper.BODY_INFORMATION_IGNITION_STATUS);
             bodyInformationObj.put(BodyInformation.KEY_IGNITION_STATUS, VehicleDataHelper.BODY_INFORMATION_IGNITION_STABLE_STATUS);
+            bodyInformationObj.put(BodyInformation.KEY_ROOF_STATUSES, VehicleDataHelper.ROOF_STATUES);
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/notifications/OnVehicleDataTests.java:323]: @kostyaBoss I think you can revert these changes, as vehicle data does not need to be changed for this PR. The corresponding PR for the JavaScript suite (https://github.com/smartdevicelink/sdl_javascript_suite/pull/354/files) did not make changes to the vehicle data.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java [hunk 9]
```diff
@@ -5,16 +5,19 @@
 import com.smartdevicelink.proxy.rpc.BodyInformation;
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java:5]: @kostyaBoss I think you can revert these changes, as vehicle data does not need to be changed for this PR. The corresponding PR for the JavaScript suite (https://github.com/smartdevicelink/sdl_javascript_suite/pull/354/files) did not make changes to the vehicle data.

**Comment by kostyaBoss** [comment]
@santhanamk I've processed the changes, please re-review the PR

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/DoorStatusTests.java [hunk 10]
```diff
@@ -34,7 +32,7 @@ public void testRpcValues() {
 
         // Valid Tests
         assertTrue(Validator.validateGrid(msg.getLocation(), location));
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/DoorStatusTests.java:34]: @kostyaBoss You can delete line 34:  `assertTrue(Validator.validateGrid(msg.getLocation(), location));`.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/DoorStatusTests.java [hunk 11]
```diff
@@ -34,7 +32,7 @@ public void testRpcValues() {
...
-        assertEquals(TestValues.MATCH, TestValues.GENERAL_DOOR_STATUS_TYPE, status);
+        assertEquals(TestValues.MATCH, TestValues.GENERAL_GRID, location);
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/DoorStatusTests.java:35]: @kostyaBoss Below line 35 you can bring back the line:

`assertEquals(TestValues.MATCH, TestValues.GENERAL_DOOR_STATUS_TYPE, status);`

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java [hunk 12]
```diff
@@ -5,16 +5,19 @@
...
 import com.smartdevicelink.proxy.rpc.ClusterModeStatus;
 import com.smartdevicelink.proxy.rpc.DeviceStatus;
+import com.smartdevicelink.proxy.rpc.DoorStatus;
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java:8]: @kostyaBoss You can revert the changes in `VehicleDataHelper.java`.

You can delete the unused import on line 8.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java [hunk 13]
```diff
@@ -5,16 +5,19 @@
...
 import com.smartdevicelink.proxy.rpc.ECallInfo;
 import com.smartdevicelink.proxy.rpc.EmergencyEvent;
 import com.smartdevicelink.proxy.rpc.FuelRange;
 import com.smartdevicelink.proxy.rpc.GPSData;
+import com.smartdevicelink.proxy.rpc.GateStatus;
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java:13]: @kostyaBoss You can delete the unused import on line 13.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java [hunk 14]
```diff
@@ -5,16 +5,19 @@
...
 import com.smartdevicelink.proxy.rpc.GearStatus;
 import com.smartdevicelink.proxy.rpc.GetVehicleDataResponse;
 import com.smartdevicelink.proxy.rpc.Grid;
 import com.smartdevicelink.proxy.rpc.HeadLampStatus;
 import com.smartdevicelink.proxy.rpc.MyKey;
 import com.smartdevicelink.proxy.rpc.OnVehicleData;
+import com.smartdevicelink.proxy.rpc.RoofStatus;
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java:20]: @kostyaBoss You can delete the unused import on line 20.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java [hunk 15]
```diff
@@ -48,6 +51,7 @@
 
 import org.json.JSONArray;
 import org.json.JSONException;
+import org.junit.Test;
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java:54]: @kostyaBoss You can delete the unused import on line 54.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java [hunk 16]
```diff
@@ -147,6 +151,10 @@
     public static final Boolean BODY_INFORMATION_REAR_LEFT_AJAR = false;
     public static final Boolean BODY_INFORMATION_REAR_RIGHT_AJAR = true;
 
+    public static final JSONArray ROOF_STATUES = TestValues.JSON_ROOF_STATUSES;
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/VehicleDataHelper.java:154]: @kostyaBoss You can delete lines 154-156.

**Review by santhanamk — COMMENTED** [review]
@kostyaBoss I have gone through the changes.  Code looks good.  I have given a few more suggestions.  Please review whenever you get the chance.

**Comment by kostyaBoss** [comment]
@santhanamk I've updated the RP, please re-review

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/DoorStatusTests.java [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        msg = new DoorStatus(TestValues.GENERAL_GRID, TestValues.GENERAL_DOOR_STATUS_TYPE);
+    }
+
+    public void testRpcValues() {
+        // Test Values
+        Grid location = msg.getLocation();
+        DoorStatusType status = msg.getStatus();
+
+        // Valid Tests
+        assertEquals(TestValues.MATCH, TestValues.GENERAL_GRID, location);
```
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/datatypes/DoorStatusTests.java:34]: @kostyaBoss You can add this line below line 34:

`assertEquals(TestValues.MATCH, TestValues.GENERAL_DOOR_STATUS_TYPE, status);`

**Review by santhanamk — COMMENTED** [review]
@kostyaBoss Code looks good.  I have suggested one more change.  Please take a look whenever you get the chance.

**Comment by kostyaBoss** [comment]
@santhanamk I've processed the changes, please, re-review

**Review by santhanamk — APPROVED** [review]
@kostyaBoss Code looks good. I have approved the PR.

Once you have tested against Core and HMI (if applicable), please put the links for those PRs. Please mark the PR as ready for review, and you can tag Livio once the testing is done on your end.

**Comment by kostyaBoss** [comment]
@bilal-alsharifi , @jordynmackool the PR is ready for the review

**Review by bilal-alsharifi — CHANGES_REQUESTED** [review]

**Comment by kostyaBoss** [comment]
@bilal-alsharifi I've processed the comments, please, re-review the PR

File: base/src/main/java/com/smartdevicelink/proxy/rpc/BodyInformation.java [hunk 18]
```diff
@@ -38,6 +38,7 @@
 import com.smartdevicelink.proxy.rpc.enums.IgnitionStatus;
 
 import java.util.Hashtable;
+import java.util.List;
 
 /**
  * The body information including power modes.
```
Comment (bilal-alsharifi) [inline base/src/main/java/com/smartdevicelink/proxy/rpc/BodyInformation.java:44]: please replace this
```
 * 			<td>SmartDeviceLink 2.0</td>
 ```

with that:
```
 * 			<td>
 * 				@since SmartDeviceLink 2.0.0
 * 				@deprecated  in SmartDeviceLink 7.1.0
 * 			</td>
 ```

 For `driverDoorAjar`, `passengerDoorAjar`, `rearLeftDoorAjar`, `rearRightDoorAjar` to indicate that they are deprecated

**Review by bilal-alsharifi — COMMENTED** [review]
@kostyaBoss finished reviewing and added one more comment.

File: base/src/main/java/com/smartdevicelink/proxy/rpc/BodyInformation.java (hunk 18 above)
Comment (kostyaBoss) [reply base/src/main/java/com/smartdevicelink/proxy/rpc/BodyInformation.java:44]: @bilal-alsharifi I've made corresponding changes, please, re-review

**Review by bilal-alsharifi — APPROVED** [review]
